### PR TITLE
Refactor the logic of assign tasks in IndexCoord

### LINKED ISSUE #6459: Use timers to assign tasks in IndexCoord
**Please state your issue using the following template and, most importantly, in English.**
Use timers to assign tasks in IndexCoord to prevent cretain tasks from failing retry due to the IndexNode crashed.

**Is your feature request related to a problem? Please describe.**
When IndexCoord's nodetasks has not recorded the task, the IndexNode cashes. I hope this task can be executed by other nodes.

**Describe the solution you'd like**
1. Use timers to assign tasks in IndexCoord, including unassigned tasks and tasks corresponding to IndexNode crashes.
2. Delete the nodetasks mapping.
3. Delete the `RegisterNode`, IndexCoord monitors ETCD to follow the online and offline of IndexNode.

**Describe alternatives you've considered**

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

### PR DESCRIPTION
**Pull request by xiaocai2333**
1. Remove RegisterNode.
2. Assign building index tasks by timetick.
3. Remove mapping of nodeID to tasks.

issue: #6427 
Signed-off-by: cai.zhang cai.zhang@zilliz.com

### CODE REVIEW

File: internal/indexnode/indexnode.go [hunk 1]
```diff
@@ -106,38 +105,13 @@ func (i *IndexNode) Init() error {
 	log.Debug("IndexNode try connect etcd success")
 	log.Debug("IndexNode start to wait for IndexCoord ready")
 
+	//TODO: IndexNode does not need to wait IndexCoord healthy.
```
Comment (czs007) [inline internal/indexnode/indexnode.go:108]: Please remove the logic of waiting for IndexCoord Ready

File: internal/indexcoord/priority_queue.go [hunk 2]
```diff
@@ -12,71 +12,60 @@
 package indexcoord
 
 import (
```
Comment (czs007) [inline internal/indexcoord/priority_queue.go:14]: revert this change

**Comment by xiaocai2333** [comment]
#6459

File: internal/indexcoord/index_coord.go [hunk 3]
```diff
@@ -542,18 +487,20 @@ func (i *IndexCoord) watchNodeLoop() {
...
 			return
 		case event := <-i.eventChan:
+			log.Debug("IndexCoord watchNodeLoop event updated")
 			switch event.EventType {
 			case sessionutil.SessionAddEvent:
 				serverID := event.Session.ServerID
 				log.Debug("IndexCoord watchNodeLoop SessionAddEvent", zap.Any("serverID", serverID))
+				err := i.nodeManager.AddNode(serverID, event.Session.Address)
+				if err != nil {
+					log.Debug("IndexCoord", zap.Any("Add IndexNode err", err))
```
Comment (czs007) [inline internal/indexcoord/index_coord.go:497]: add serverID, Address to the log

File: internal/indexcoord/meta_table.go [hunk 4]
```diff
@@ -361,28 +361,32 @@ func (mt *metaTable) GetIndexMeta(indexBuildID UniqueID) Meta {
...
+			metas = append(metas, meta)
+			continue
+		}
+		if meta.indexMeta.State == commonpb.IndexState_Finished || meta.indexMeta.State == commonpb.IndexState_Failed {
+			continue
+		}
 		alive := false
 		for _, serverID := range nodeIDs {
 			if meta.indexMeta.NodeID == serverID {
 				alive = true
```
Comment (czs007) [inline internal/indexcoord/meta_table.go:381]: you can break early

File: internal/indexcoord/meta_table.go [hunk 5]
```diff
@@ -477,67 +481,16 @@ func (mt *metaTable) LoadMetaFromETCD(indexBuildID int64, revision int64) bool {
...
-		return
-	}
-	indexBuildIDs = append(indexBuildIDs, indexBuildID)
-	nt.nodeID2Tasks[serverID] = indexBuildIDs
-}
-
-func (nt *nodeTasks) finishTask(indexBuildID UniqueID) {
-	nt.lock.Lock()
-	defer nt.lock.Unlock()
+func (mt *metaTable) GetPriorityForNodeID() map[UniqueID]int {
```
Comment (czs007) [inline internal/indexcoord/meta_table.go:484]: please rename to GetNodeTaskStats

File: internal/indexcoord/meta_table.go [hunk 6]
```diff
@@ -477,67 +481,16 @@ func (mt *metaTable) LoadMetaFromETCD(indexBuildID int64, revision int64) bool {
...
+	mt.lock.Lock()
```
Comment (czs007) [inline internal/indexcoord/meta_table.go:485]: Use Rlock and RUnlock

File: internal/indexcoord/priority_queue.go [hunk 7]
```diff
@@ -137,21 +132,21 @@ func (pq *PriorityQueue) Peek() interface{} {
 }
 
 // PeekClient picks an IndexNode with the lowest load.
-func (pq *PriorityQueue) PeekClient() (UniqueID, types.IndexNode) {
+func (pq *PriorityQueue) PeekClient() UniqueID {
```
Comment (czs007) [inline internal/indexcoord/priority_queue.go:135]: replace Peek() and return UniqueID

File: internal/indexcoord/priority_queue.go [hunk 8]
```diff
@@ -71,12 +66,12 @@ func (pq *PriorityQueue) Pop() interface{} {
 	return item
 }
 
-func (pq *PriorityQueue) CheckAddressExist(addr *commonpb.Address) bool {
+func (pq *PriorityQueue) CheckNodeIDExist(nodeID UniqueID) bool {
```
Comment (czs007) [inline internal/indexcoord/priority_queue.go:69]: func (pq *PriorityQueue) CheckExist(key UniqueID) bool {

File: internal/indexcoord/meta_table.go [hunk 9]
```diff
@@ -361,28 +361,32 @@ func (mt *metaTable) GetIndexMeta(indexBuildID UniqueID) Meta {
 	return meta
 }
 
-func (mt *metaTable) GetUnassignedTasks(nodeIDs []int64) [][]UniqueID {
-	var tasks [][]UniqueID
-	var indexBuildIDs []UniqueID
+func (mt *metaTable) GetUnassignedTasks(nodeIDs []int64) []Meta {
```
Comment (czs007) [inline internal/indexcoord/meta_table.go:364]: better to rename nodeIDs to onlineNodeIDs

File: internal/indexcoord/index_coord.go [hunk 10]
```diff
@@ -45,13 +45,11 @@ import (
 const (
 	reqTimeoutInterval = time.Second * 10
 	durationInterval   = time.Second * 10
-	recycleIndexLimit  = 20
+	assignTasksLimit   = 20
```
Comment (czs007) [inline internal/indexcoord/index_coord.go:48]: Renamed to a neutral word

File: internal/indexcoord/index_coord.go [hunk 11]
```diff
@@ -596,38 +543,70 @@ func (i *IndexCoord) watchMetaLoop() {
...
+		case <-ctx.Done():
+			return
+		case <-timeTicker.C:
+			sessions, _, err := i.session.GetSessions(typeutil.IndexNodeRole)
+			if err != nil {
+				log.Debug("IndexCoord assignTaskLoop", zap.Any("GetSessions error", err))
+			}
+			var serverIDs []int64
+			for _, session := range sessions {
+				serverIDs = append(serverIDs, session.ServerID)
```
Comment (czs007) [inline internal/indexcoord/index_coord.go:566]: If serverIDs is empty, you can skip the following logic

File: internal/indexcoord/index_coord.go [hunk 12]
```diff
@@ -596,38 +543,70 @@ func (i *IndexCoord) watchMetaLoop() {
...
-				Port: port,
-			},
-			NodeID: session.ServerID,
-		}
-		if err = i.addNode(session.ServerID, req); err != nil {
-			log.Debug("IndexCoord", zap.Any("IndexCoord start find node fatal, err = ", err))
+	defer cancel()
+	defer i.loopWg.Done()
+
+	timeTicker := time.NewTicker(durationInterval)
```
Comment (czs007) [inline internal/indexcoord/index_coord.go:552]: 10 seconds is too long.
please change to 3 seconds.
In the future, a calculation formula can be designed to dynamically modify the number of tasks assigned each time and the timer interval according to the total task load, the rate at which nodes complete tasks, the number of nodes, etc.

Comment (czs007) [reply internal/indexcoord/index_coord.go:552]: In the future, it is also necessary to add information about the opening time of the task to the meta. You can make scheduling based on this information to minimize the average waiting time of the task.

Comment (xiaocai2333) [reply internal/indexcoord/index_coord.go]: Good suggestion.
